### ISSUE TITLE
FE_Nothing ends up dominating when in an FE_System

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#1496]
@davydden posted a testcase as part of #1424 (now hidden by follow-up commits) that compared a Q2xQ1 against a Q1xNothing. We should get hanging node constraints where the Q2 and the Q1 come together, but no additional constraints for the Q1-vs-Nothing. But we do. That's wrong.

The testcase that goes with this (and that currently produces the wrong answer) is `hp/face_domination_02`, as introduced by #1495. Someone will need to debug `make_hanging_node_constraints` to see what is going wrong. I fear that the problem is that we end up with this because the two vector components have to be treated differently but we only have a single matrix to do this.

**Comment by bangerth** [comment]
![2015-08-31-173257](https://cloud.githubusercontent.com/assets/7504421/9592183/6aeaa25c-5006-11e5-99ee-4921f0b236e3.jpg)

As shown, we should get (and do get) the constraint `x_7=1/2(x_13+x_11)`. But we also do get `x_5=0` and `x_2=0`, and this is wrong.

**Comment by davydden** [comment]
>  I fear that the problem is that we end up with this because the two vector components have to be treated differently but we only have a single matrix to do this.

agreed. Supposedly one would need to introduce a special treatment for FESystem (vector FE which can be broken into a separate FEs) but use the current logic for other vector FEs like FE_RaviartThomas.

p.s. here's a link to the original comment https://github.com/dealii/dealii/pull/1424#discussion_r37973319

**Comment by davydden** [comment]
i hit that issue again while trying to implement something. It took a while to recall what's happening and why. A workaround could be to delete those constraints from constraint matrix, but this is dirty...

The issue is indeed related to the fact that we have a single matrix to treat/introduce constraints.
In scenarios depicted above we end up with 
`case FiniteElementDomination::either_element_can_dominate` (around line [1173](https://github.com/dealii/dealii/blob/master/source/dofs/dof_tools_constraints.cc#L1173)).
Consequently we constrain DoFs on child faces to those on the parent face by calling, at the end of the day,
```
ensure_existence_of_subface_matrix
(cell->get_fe(),
subface->get_fe(subface_fe_index),
c,
subface_interpolation_matrices
[cell->active_fe_index()][subface_fe_index][c]);
```
and
```
// Add constraints to global constraint matrix.
 filter_constraints 
(master_dofs,
slave_dofs,
*(subface_interpolation_matrices
[cell->active_fe_index()][subface_fe_index][c]),
constraints);
```

There are several directions to explore while looking for a solution (at least for that `simple` case where it is clear who dominates and we don't need to search for a dominating FE inside FECollection):

1. Everywhere when we have `cell->get_fe()` we need to check whether or not this is a FESystem or not. And if it is FESystem, then we go another `if-else` branch where things are treated separately for each component. I feel this is too cumbersome and would require introduciton of extra data structures to hold projection matrices etc.

2.  I think the best way is to introduce extra argument to `filter_constraints` to specify exclusion range for slave DoFs. AFAIK, in FESystem all the projection matrices are block-structured, so we can do an extra run through each base FE to see if there is `no_requirements` for that block. If that's the case the corresponding entries can be neglected when populating `ConstraintMatrix`.

**Comment by bangerth** [comment]
I'll see whether I can get this done before the next release. But I think the issue is not trivial because there are multiple ways to create a vector FE. FESystem is only one, but you could in principle write your own implementation that works exactly like FESystem but is called differently; in those cases, the implementation of constraints should continue to work the same way it always did, even though the element "is not an FESystem". In other words, I can think of many complications :-(

**Comment by davydden** [comment]
true. The approach I had in mind won't work for a general case. Well, then i guess one has shift the logic into component-by-component basis, introduce interfaces for domination for a given component, projection matrices, etc.

**Comment by bangerth** [comment]
Yes, though you can't separate the components of, say, a Raviart-Thomas element. You'll have to work with "blocks". Which of course also yields the question of how to form constraints between, say, a Raviart-Thomas element and a Q(p)^dim...

**Comment by davydden** [comment]
@bangerth i suppose we postpone it until `9.0`? If so, please change the milestone.

**Comment by bangerth** [comment]
Yes. I just don't seem to have time this week or next :-(

**Comment by davydden** [comment]
@bangerth my student and I hit this issue again in https://github.com/dealii/dealii/pull/6277 . So far we have a work-around but it might be that it's better to address the issue as I am not sure that the work-around is robust. I checked the `make_hanging_node_constraints` recently, but I am not anymore sure what's the right course of action here. If we assume that FESystem can only be combined with FESystem of the same `size`, i.e. `Q2xQ1` vs `Q1xQ2` as opposed to `Q1xQ2xQ3`), then one could change the logic a bit and go through all elements in FESystem one-by-one. It is also related to the way DoFs are distributed on faces, if I am not mistaken currently two neighbouring elements `Q1xQ1xQ1xQ2` and `Q1xQ1xQ1xQ1` will have all different DoFs. Whereas in reality we only need them to be different for the last component (i.e. `Q2` vs `Q1`). Perhaps we can discuss it in mode detail in Italy.

**Comment by marcfehling** [comment]
Following this issue, I stumbled over the test `fe/fe_nothing` ([code](https://github.com/dealii/dealii/blob/master/tests/fe/fe_nothing.cc), [output](https://github.com/dealii/dealii/blob/master/tests/fe/fe_nothing.output)) which depicts the exact same domain and yields the same constraints.

Just for your interest.

**Comment by kronbichler** [comment]
Despite the reoccurring interest, this will have to wait until the next release, as there does not seem to be a complete solution yet.

**Comment by marcfehling** [comment]
If I understood correctly, we add constraints for every row in face interpolation matrices (that do not correspond to identities). Whenever a pure `FENothing` element is involved, these matrices are empty with `m x 0` or `0 x n` entries.

As part of `FESystem` this is no longer the case. From my understanding, there will simply be rows with zeros that correspond to dofs of the `FENothing` counterpart on the other side of the face.

Here, we will simply add the `0` constraint to these lines as described above.
https://github.com/dealii/dealii/blob/1fe592ad8fb97c57018eccaf3dfb74c6b3201376/source/dofs/dof_tools_constraints.cc#L542-L550

---

Would it work if we simply ignore rows that entirely consist of zeros, i.e. skip this code snippet entirely when the row entirely consists of zeros? I'll give it a try and work on a patch.

**Comment by marcfehling** [comment]
Hmm, this problem seems to persist in constraints with `FE_Enriched` as well since its implementation is based on `FE_System`. For example, the `= 0` constraints in
https://github.com/dealii/dealii/blob/1fe592ad8fb97c57018eccaf3dfb74c6b3201376/tests/fe/fe_enriched_08.with_cxx14%3Don.output#L3-L4

My patch will remove these constraints as well.

**Comment by marcfehling** [comment]
I've been working on this, but will most likely not find a fix for this. We now have a multitude of tests that reflect the current state of this issue, which will hopefully make it easier to find a fix for the 10.0 release.

**Comment by peterrum** [comment]
@marcfehling Are you planing to work on `FE_Nothing` before the release? Else we could change the milestone.

**Comment by marcfehling** [comment]
It's on a high priority on my list, but I'm almost sure that it won't make it in the next release. Let's adjust the milestone.
